File: Examples/Comparison/clustal_muscle_compare.py
from pathlib import Path
from subprocess import run
from tempfile import TemporaryDirectory
from time import perf_counter
import os
import sys

from Bio import AlignIO
from Bio.Align import MultipleSeqAlignment
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord


def build_sequences(record_count=8, width=180):
    template = ("ACGT" * ((width + 3) // 4))[:width]
    bases = "ACGT"
    records = []
    for record_index in range(1, record_count + 1):
        chars = list(template)
        for position in range(1, width + 1):
            if (record_index + position) % 19 == 0:
                chars[position - 1] = bases[(record_index + position - 2) % len(bases)]
            elif (record_index + position) % 13 == 0:
                chars[position - 1] = bases[(record_index + position) % len(bases)]
        records.append(SeqRecord(Seq("".join(chars)), id=f"seq{record_index}", description="synthetic_sequence"))
    return records


def time_call(func):
    start = perf_counter()
    result = func()
    return (perf_counter() - start) * 1000.0, result


def resolve_binary(candidate):
    candidates = [candidate]
    conda_prefix = os.environ.get("CONDA_PREFIX")
    if conda_prefix:
        candidates.append(str(Path(conda_prefix) / "bin" / candidate))
    home = Path.home()
    for root in [home / "miniconda3" / "envs" / "general", home / ".conda" / "envs" / "general", home / "anaconda3" / "envs" / "general"]:
        candidates.append(str(root / "bin" / candidate))
    for path in candidates:
        if Path(path).is_file():
            return path
    raise FileNotFoundError(candidate)


def write_fasta(records, path):
    AlignIO.write(MultipleSeqAlignment(records), str(path), "fasta")


def run_clustalo(binary, input_path, output_path):
    run([binary, "--infile", str(input_path), "--outfile", str(output_path), "--outfmt", "fasta", "--force"], check=True, capture_output=True)


def run_muscle(binary, input_path, output_path):
    attempts = [
        [binary, "-align", str(input_path), "-output", str(output_path)],
        [binary, "-in", str(input_path), "-out", str(output_path)],
    ]
    last_error = None
    for command in attempts:
        try:
            run(command, check=True, capture_output=True)
            return
        except Exception as exc:
            last_error = exc
    raise last_error


def roundtrip(binary, runner, records, suffix, temp_dir):
    input_path = Path(temp_dir) / f"input{suffix}.fasta"
    output_path = Path(temp_dir) / f"output{suffix}.fasta"
    write_fasta(records, input_path)
    runner(binary, input_path, output_path)
    return AlignIO.read(str(output_path), "fasta")


def assert_alignment(alignment, expected_count):
    assert len(alignment) == expected_count
    assert alignment.get_alignment_length() > 0
    assert all(str(record.seq) for record in alignment)


def main():
    records = build_sequences()
    clustalo = resolve_binary("clustalo")
    muscle = resolve_binary("muscle")

    with TemporaryDirectory() as temp_dir:
        clustal_ms, clustal_alignment = time_call(lambda: roundtrip(clustalo, run_clustalo, records, "clustal", temp_dir))
        muscle_ms, muscle_alignment = time_call(lambda: roundtrip(muscle, run_muscle, records, "muscle", temp_dir))

        assert_alignment(clustal_alignment, len(records))
        assert_alignment(muscle_alignment, len(records))

        print(f"python_clustal_msa_ms={clustal_ms:.4f}")
        print(f"python_muscle_msa_ms={muscle_ms:.4f}")
        print(f"python_external_msa_records={len(clustal_alignment)}")
        print(f"python_external_msa_columns={clustal_alignment.get_alignment_length()}")
        print("python_clustal_msa_ok=1")
        print("python_muscle_msa_ok=1")


if __name__ == "__main__":
    main()

File: Examples/Comparison/clustal_muscle_compare.jl
pushfirst!(LOAD_PATH, abspath(joinpath(@__DIR__, "..", "..")))

using BioToolkit
include(joinpath(@__DIR__, "..", "..", "scripts", "benchmark_helpers.jl"))

function build_sequences(record_count::Integer=8, width::Integer=180)
    template = repeat("ACGT", cld(width, 4))[1:width]
    bases = ('A', 'C', 'G', 'T')
    records = SeqRecordLite[]

    for record_index in 1:record_count
        chars = collect(template)
        for position in 1:width
            if mod(record_index + position, 19) == 0
                chars[position] = bases[mod(record_index + position - 2, length(bases)) + 1]
            elseif mod(record_index + position, 13) == 0
                chars[position] = bases[mod(record_index + position, length(bases)) + 1]
            end
        end
        sequence = String(chars)
        push!(records, SeqRecordLite(sequence; identifier="seq$(record_index)", description="synthetic_sequence"))
    end

    return records
end

function resolve_binary(candidate::AbstractString)
    resolved = Sys.which(candidate)
    resolved !== nothing && return resolved

    home = homedir()
    for root in (
        get(ENV, "CONDA_PREFIX", ""),
        joinpath(home, "miniconda3", "envs", "general"),
        joinpath(home, ".conda", "envs", "general"),
        joinpath(home, "anaconda3", "envs", "general"),
    )
        isempty(root) && continue
        path = joinpath(root, "bin", candidate)
        isfile(path) && return path
    end

    throw(ArgumentError("unable to locate binary: $(candidate)"))
end

function assert_alignment(alignment, expected_count::Integer)
    @assert length(alignment) == expected_count
    @assert get_alignment_length(alignment) > 0
    @assert all(!isempty(record.sequence) for record in alignment.records)
    return nothing
end

function main()
    records = build_sequences()
    clustalo = resolve_binary("clustalo")
    muscle = resolve_binary("muscle")
    temp_dir = mktempdir()
    try
        julia_clustal_ms, julia_clustal = elapsed_ms() do
            BioToolkit.clustal_msa(records; executable=clustalo, output_format="fasta")
        end

        julia_muscle_ms, julia_muscle = elapsed_ms() do
            BioToolkit.muscle_msa(records; executable=muscle, output_format="fasta")
        end

        assert_alignment(julia_clustal, length(records))
        assert_alignment(julia_muscle, length(records))

        python_output = read(`conda run -n general --no-capture-output python $(joinpath(@__DIR__, "clustal_muscle_compare.py")) $(clustalo) $(muscle)`, String)

        python_clustal_ms = parse(Float64, match(r"python_clustal_msa_ms=([0-9.]+)", python_output).captures[1])
        python_muscle_ms = parse(Float64, match(r"python_muscle_msa_ms=([0-9.]+)", python_output).captures[1])
        python_records = parse(Int, match(r"python_external_msa_records=(\d+)", python_output).captures[1])
        python_columns = parse(Int, match(r"python_external_msa_columns=(\d+)", python_output).captures[1])

        @assert python_records == length(records)
        @assert python_columns > 0
        @assert occursin("python_clustal_msa_ok=1", python_output)
        @assert occursin("python_muscle_msa_ok=1", python_output)

        println("External MSA benchmark")
        println("  record_count=", length(records))
        println("  sequence_length=180")
        println("  clustal_binary=", clustalo)
        println("  muscle_binary=", muscle)
        println("  julia_clustal_msa_ms=", round(julia_clustal_ms, digits=4))
        println("  julia_muscle_msa_ms=", round(julia_muscle_ms, digits=4))
        println("  julia_external_msa_records=", length(julia_clustal))
        println("  julia_external_msa_columns=", get_alignment_length(julia_clustal))
        print(python_output)
    finally
        rm(temp_dir; recursive=true, force=true)
    end
end

main()
